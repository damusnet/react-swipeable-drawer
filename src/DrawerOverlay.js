/* eslint jsx-a11y/click-events-have-key-events: 0 */
import React from "react";
import PropTypes from "prop-types";

const transform = ({ position, swiping, open }) => {
  switch (position) {
    case "top":
      return {
        top: 0,
        left: 0,
        right: 0,
        height: swiping || open ? "100%" : "20px",
      };
    case "right":
      return {
        right: 0,
        top: 0,
        bottom: 0,
        width: swiping || open ? "100%" : "20px",
      };
    case "bottom":
      return {
        bottom: 0,
        left: 0,
        right: 0,
        height: swiping || open ? "100%" : "20px",
      };
    case "left":
    default:
      return {
        left: 0,
        top: 0,
        bottom: 0,
        width: swiping || open ? "100%" : "20px",
      };
  }
};

const transition = ({ swiping, open }) => {
  if (swiping) {
    return "";
  } else if (open) {
    return "background-color .2s ease-in-out, width 0s 0s, height 0s 0s";
  }
  return "background-color .2s ease-in-out, width 0s .2s, height 0s 0s";
};

const DrawerOverlay = ({
  position,
  zIndex,
  open,
  swiping,
  translation,
  toggleDrawer,
  handleTouchStart,
  handleTouchMove,
  handleTouchEnd,
}) => (
  <div
    className="DrawerOverlay"
    onClick={open ? toggleDrawer : null}
    onTouchStart={handleTouchStart}
    onTouchMove={handleTouchMove(100)}
    onTouchEnd={handleTouchEnd}
    style={{
      position: "fixed",
      zIndex,
      backgroundColor: `rgba(0,0,0,${0.6 * translation / 100})`,
      transition: transition({ swiping, open }),
      ...transform({ position, swiping, open }),
    }}
  />
);

export default DrawerOverlay;

DrawerOverlay.propTypes = {
  position: PropTypes.oneOf(["left", "right", "top", "bottom"]).isRequired,
  zIndex: PropTypes.number.isRequired,
  open: PropTypes.bool.isRequired,
  swiping: PropTypes.bool.isRequired,
  translation: PropTypes.number.isRequired,
  toggleDrawer: PropTypes.func.isRequired,
  handleTouchStart: PropTypes.func.isRequired,
  handleTouchMove: PropTypes.func.isRequired,
  handleTouchEnd: PropTypes.func.isRequired,
};
