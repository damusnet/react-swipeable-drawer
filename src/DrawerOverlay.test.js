import React from "react";
import renderer from "react-test-renderer";

import DrawerOverlay from "./DrawerOverlay";

const noop = () => () => {};
const positions = ["left", "right", "top", "bottom"];

describe("<DrawerOverlay />", () => {
  positions.forEach(position => {
    it(`renders correctly when closed ${position}`, () => {
      const component = renderer.create(
        <DrawerOverlay
          position={position}
          open={false}
          swiping={false}
          translation={0}
          toggleDrawer={noop}
          handleTouchStart={noop}
          handleTouchMove={noop}
          handleTouchEnd={noop}
          style={{ zIndex: 1 }}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders correctly when open", () => {
      const component = renderer.create(
        <DrawerOverlay
          position={position}
          open
          swiping={false}
          translation={100}
          toggleDrawer={noop}
          handleTouchStart={noop}
          handleTouchMove={noop}
          handleTouchEnd={noop}
          style={{ zIndex: 1 }}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it("renders correctly when swiping", () => {
    const component = renderer.create(
      <DrawerOverlay
        position="left"
        open={false}
        swiping
        translation={50}
        toggleDrawer={noop}
        handleTouchStart={noop}
        handleTouchMove={noop}
        handleTouchEnd={noop}
        style={{ zIndex: 1 }}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
