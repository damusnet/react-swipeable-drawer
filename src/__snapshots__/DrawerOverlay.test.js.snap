// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerOverlay /> renders correctly when closed bottom 1`] = `
<div
  className="DrawerOverlay"
  onClick={null}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "bottom": 0,
      "height": "20px",
      "left": 0,
      "position": "fixed",
      "right": 0,
      "transition": "background-color .2s ease-in-out, width 0s .2s, height 0s 0s",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when closed left 1`] = `
<div
  className="DrawerOverlay"
  onClick={null}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s .2s, height 0s 0s",
      "width": "20px",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when closed right 1`] = `
<div
  className="DrawerOverlay"
  onClick={null}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "bottom": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s .2s, height 0s 0s",
      "width": "20px",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when closed top 1`] = `
<div
  className="DrawerOverlay"
  onClick={null}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "height": "20px",
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s .2s, height 0s 0s",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when open 1`] = `
<div
  className="DrawerOverlay"
  onClick={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.6)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s 0s, height 0s 0s",
      "width": "100%",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when open 2`] = `
<div
  className="DrawerOverlay"
  onClick={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.6)",
      "bottom": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s 0s, height 0s 0s",
      "width": "100%",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when open 3`] = `
<div
  className="DrawerOverlay"
  onClick={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.6)",
      "height": "100%",
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "transition": "background-color .2s ease-in-out, width 0s 0s, height 0s 0s",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when open 4`] = `
<div
  className="DrawerOverlay"
  onClick={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.6)",
      "bottom": 0,
      "height": "100%",
      "left": 0,
      "position": "fixed",
      "right": 0,
      "transition": "background-color .2s ease-in-out, width 0s 0s, height 0s 0s",
      "zIndex": 1,
    }
  }
/>
`;

exports[`<DrawerOverlay /> renders correctly when swiping 1`] = `
<div
  className="DrawerOverlay"
  onClick={null}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.3)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "top": 0,
      "transition": "",
      "width": "100%",
      "zIndex": 1,
    }
  }
/>
`;
